<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head><title>R: some description</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<link rel="stylesheet" type="text/css" href="R.css" />
</head><body>
<h1> some description
<img class="toplogo" src="../../../doc/html/logo.jpg" alt="[R logo]" />
</h1>
<hr/>
<div style="text-align: center;">
<a href="../../../doc/html/packages.html"><img class="arrow" src="../../../doc/html/left.jpg" alt="[Up]" /></a>
<a href="../../../doc/html/index.html"><img class="arrow" src="../../../doc/html/up.jpg" alt="[Top]" /></a>
</div><h2>Documentation for package &lsquo;spp&rsquo; version 1.13</h2>

<ul><li><a href="../DESCRIPTION">DESCRIPTION file</a>.</li>
</ul>

<h2>Help Pages</h2>


<table width="100%">
<tr><td style="width: 25%;"><a href="spp-package.html">spp-package</a></td>
<td>ChIP-seq (Solexa) Processing Pipeline</td></tr>
<tr><td style="width: 25%;"><a href="add.broad.peak.regions.html">add.broad.peak.regions</a></td>
<td>Calculate chromosome-wide profiles of smoothed tag density</td></tr>
<tr><td style="width: 25%;"><a href="find.binding.positions.html">find.binding.positions</a></td>
<td>Determine significant point protein binding positions (peaks)</td></tr>
<tr><td style="width: 25%;"><a href="get.binding.characteristics.html">get.binding.characteristics</a></td>
<td>Calculate characteristics of observed DNA-binding signal from cross-correlation profiles</td></tr>
<tr><td style="width: 25%;"><a href="get.broad.enrichment.clusters.html">get.broad.enrichment.clusters</a></td>
<td>Determine broad clusters of enrichment</td></tr>
<tr><td style="width: 25%;"><a href="get.conservative.fold.enrichment.profile.html">get.conservative.fold.enrichment.profile</a></td>
<td>Estimate minimal fold enrichment/depletion along the chromosomes</td></tr>
<tr><td style="width: 25%;"><a href="get.mser.html">get.mser</a></td>
<td>Calculate minimal saturated enrichment fold ratio</td></tr>
<tr><td style="width: 25%;"><a href="get.mser.interpolation.html">get.mser.interpolation</a></td>
<td>Interpolate MSER dependency on the tag count</td></tr>
<tr><td style="width: 25%;"><a href="get.smoothed.enrichment.mle.html">get.smoothed.enrichment.mle</a></td>
<td>Calculate chromosome-wide profiles of smoothed enrichment estimate</td></tr>
<tr><td style="width: 25%;"><a href="get.smoothed.tag.density.html">get.smoothed.tag.density</a></td>
<td>Calculate chromosome-wide profiles of smoothed tag density</td></tr>
<tr><td style="width: 25%;"><a href="output.binding.results.html">output.binding.results</a></td>
<td>Write out determined binding peaks into a text file table</td></tr>
<tr><td style="width: 25%;"><a href="read.bam.tags.html">read.bam.tags</a></td>
<td>Read BAM alignment file</td></tr>
<tr><td style="width: 25%;"><a href="read.bin.maqmap.tags.html">read.bin.maqmap.tags</a></td>
<td>Read MAQ binary alignment map file</td></tr>
<tr><td style="width: 25%;"><a href="read.bowtie.tags.html">read.bowtie.tags</a></td>
<td>Read bowtie text alignment output file</td></tr>
<tr><td style="width: 25%;"><a href="read.eland.tags.html">read.eland.tags</a></td>
<td>Read eland output file</td></tr>
<tr><td style="width: 25%;"><a href="read.maqmap.tags.html">read.maqmap.tags</a></td>
<td>Read MAQ text alignment output file</td></tr>
<tr><td style="width: 25%;"><a href="read.meland.tags.html">read.meland.tags</a></td>
<td>Read modified BED tag alignment file that contains variable match length information</td></tr>
<tr><td style="width: 25%;"><a href="remove.local.tag.anomalies.html">remove.local.tag.anomalies</a></td>
<td>Restrict or remove positions with too many tags relative to local background.</td></tr>
<tr><td style="width: 25%;"><a href="select.informative.tags.html">select.informative.tags</a></td>
<td>Choose informative tags</td></tr>
<tr><td style="width: 25%;"><a href="spp-package.html">spp</a></td>
<td>ChIP-seq (Solexa) Processing Pipeline</td></tr>
<tr><td style="width: 25%;"><a href="write.broadpeak.info.html">write.broadpeak.info</a></td>
<td>Write out determined broad enrichment regions using broadPeak format</td></tr>
<tr><td style="width: 25%;"><a href="write.narrowpeak.binding.html">write.narrowpeak.binding</a></td>
<td>Write out determined binding peaks using narrowPeak format</td></tr>
<tr><td style="width: 25%;"><a href="writewig.html">writewig</a></td>
<td>A function to save a list of chromosome-wise x/y data frames into a WIG file format.</td></tr>
</table>
</body></html>
